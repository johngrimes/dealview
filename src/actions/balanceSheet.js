// @flow

import { putObject, getObject } from 'db/db'
import type { BalanceSheetOverTime } from 'types/balanceSheet'
import type { AssetMap } from 'types/assets/asset'
import type { LiabilityMap } from 'types/liabilities/liability'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'BalanceSheet'
const theOneAndOnlyKey = 'BalanceSheet'

type CalculateBalanceSheetTask = {
  task: 'CALCULATE_BALANCE_SHEET',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}

type LoadBalanceSheetRequestAction = {
  type: 'LOAD_BALANCE_SHEET_REQUEST'
}
type LoadBalanceSheetSuccessAction = {
  type: 'LOAD_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime
}
type LoadBalanceSheetFailureAction = {
  type: 'LOAD_BALANCE_SHEET_FAILURE',
  error: string|null
}

type UpdateBalanceSheetRequestAction = {
  type: 'UPDATE_BALANCE_SHEET_REQUEST',
  assets: AssetMap,
  liabilities: LiabilityMap,
  startDate: string,
  endDate: string,
}
type UpdateBalanceSheetSuccessAction = {
  type: 'UPDATE_BALANCE_SHEET_SUCCESS',
  balanceSheet: BalanceSheetOverTime,
}
type UpdateBalanceSheetFailureAction = {
  type: 'UPDATE_BALANCE_SHEET_FAILURE',
  error: string|null,
}

type InvalidateBalanceSheetAction = {
  type: 'INVALIDATE_BALANCE_SHEET'
}

export type BalanceSheetAction = LoadBalanceSheetRequestAction
                               | LoadBalanceSheetSuccessAction
                               | LoadBalanceSheetFailureAction
                               | UpdateBalanceSheetRequestAction
                               | UpdateBalanceSheetSuccessAction
                               | UpdateBalanceSheetFailureAction
                               | InvalidateBalanceSheetAction

export const calculateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): CalculateBalanceSheetTask => {
  return {
    task: 'CALCULATE_BALANCE_SHEET',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const loadBalanceSheetRequest = (): LoadBalanceSheetRequestAction => {
  return { type: 'LOAD_BALANCE_SHEET_REQUEST' }
}

export const loadBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): LoadBalanceSheetSuccessAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const loadBalanceSheetFailure = (error: string|null): LoadBalanceSheetFailureAction => {
  return {
    type: 'LOAD_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const updateBalanceSheetRequest = (assets: AssetMap, liabilities: LiabilityMap, startDate: string, endDate: string): UpdateBalanceSheetRequestAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_REQUEST',
    assets,
    liabilities,
    startDate,
    endDate,
  }
}

export const updateBalanceSheetSuccess = (balanceSheet: BalanceSheetOverTime): UpdateBalanceSheetSuccessAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_SUCCESS',
    balanceSheet,
  }
}

export const updateBalanceSheetFailure = (error: string|null): UpdateBalanceSheetFailureAction => {
  return {
    type: 'UPDATE_BALANCE_SHEET_FAILURE',
    error,
  }
}

export const invalidateBalanceSheet = (): InvalidateBalanceSheetAction => {
  return { type: 'INVALIDATE_BALANCE_SHEET' }
}

export const updateBalanceSheet = (assets: AssetMap, liabilities: LiabilityMap,
                                   startDate: string, endDate: string): Thunk => {
  return dispatch => {
    dispatch(updateBalanceSheetRequest(assets, liabilities, startDate, endDate))
    return new Promise(async (resolve, reject) => {
      try {
        const worker = new Worker('/static/js/worker.js', { name: 'Balance Sheet Worker' })
        worker.onmessage = async event => {
          console.log('Balance sheet calc returned:', event.data)
          // flow-ignore
          const saved = await putObject(objectStore, event.data, theOneAndOnlyKey)
          dispatch(updateBalanceSheetSuccess(saved))
          resolve(saved)
        }
        worker.postMessage([ assets, liabilities, startDate, endDate ])
      } catch (error) {
        dispatch(updateBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}

export const loadBalanceSheet = (): Thunk => {
  return dispatch => {
    dispatch(loadBalanceSheetRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const balanceSheet = await getObject(objectStore, theOneAndOnlyKey)
        dispatch(loadBalanceSheetSuccess(balanceSheet))
        resolve(balanceSheet)
      } catch (error) {
        dispatch(loadBalanceSheetFailure(error))
        reject(error)
      }
    })
  }
}
