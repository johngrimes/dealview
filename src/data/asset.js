import moment from 'moment'

export const getValuationAtDate = (asset, date) => {
  const now = moment().startOf('day')
  const startDate = asset.purchaseDate || now
  if (
    date.isBefore(startDate) ||
    (asset.saleDate && date.isAfter(asset.saleDate))
  ) {
    return 0
  }
  const daysSinceStart = Math.floor(
    moment.duration(date.diff(startDate)).asDays()
  )
  const startValue = asset.purchaseAmount
    ? parseInt(asset.purchaseAmount, 10)
    : parseInt(asset.currentValue, 10)
  const forecastGrowth = parseInt(asset.forecastGrowth, 10) / 10000 / 365
  return Math.round(startValue * Math.pow(1 + forecastGrowth, daysSinceStart))
}
