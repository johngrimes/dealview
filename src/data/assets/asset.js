// @flow

export type Asset = {
  type: 'RealEstate',
  instanceId: string,
  name: string,
  lastValuation?: number
}

export type AssetWithId = Asset & { id: string }

export type AssetMap = { [id: string]: AssetWithId }
