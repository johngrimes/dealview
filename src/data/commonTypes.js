export const AddressDefaults = {
  line1: '',
  line2: '',
  line3: '',
  locality: '',
  state: '',
  postcode: '',
}

export const DateDisplayFormat = 'D MMM YYYY'
export const DateStorageFormat = 'YYYY-MM-DD'

export const parseNumber = string =>
  typeof string === 'string' ? parseInt(string) : undefined

export const formatNumber = number => {
  if (typeof number !== 'number') return undefined
  const numString = number.toString()
  return [...numString].reduce((prev, curr, i, arr) => {
    return arr.length - i - 1 !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const formatDollars = number =>
  typeof number === 'number' ? '$' + formatNumber(number) : undefined

export const unformatNumber = string => {
  return typeof string === 'string'
    ? parseInt(string.replace(/,/g, ''), 10)
    : undefined
}

export const unformatDollars = string => {
  return typeof string === 'string'
    ? unformatNumber(string.replace(/\$/g, ''))
    : undefined
}
