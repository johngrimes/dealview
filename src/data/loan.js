import moment from 'moment'

export const LoanDefaults = {
  interestRate: 400,
  compoundingPeriod: 'daily',
  repaymentType: 'principalAndInterest',
}

export const loanToLiability = loan => {
  const liability = {}
  liability.type = 'Loan'
  liability.name = loan.name
  liability.startDate = loan.startDate
  liability.endDate = loan.endDate
  if (typeof loan.id === 'string') {
    liability.id = loan.id
  }
  return liability
}

export const getBalanceAtDate = (loan, date) => {
  if (typeof loan !== 'object') {
    throw new TypeError(`getBalanceAtDate: loan (${loan}) is not an object`)
  }
  if (!(moment.isMoment(date) && date.isValid())) {
    throw new TypeError(
      `getBalanceAtDate: date (${date}) is not a valid Moment`
    )
  }
  const now = moment().startOf('day')
  const startDate = loan.startDate ? moment(loan.startDate) : now
  const endDate = loan.endDate ? moment(loan.endDate) : null
  if (date.isBefore(startDate) || (endDate && date.isAfter(endDate))) {
    return 0
  }
  const daysSinceStart = Math.floor(
    moment.duration(date.diff(startDate)).asDays()
  )
  const daysToLoanEnd = Math.floor(
    moment.duration(endDate.diff(startDate)).asDays()
  )
  const startValue = loan.principal
    ? parseInt(loan.principal, 10)
    : parseInt(loan.currentBalance, 10)
  const interestRate = parseInt(loan.interestRate, 10) / 10000 / 365
  const repayment =
    interestRate *
    startValue /
    (1 - Math.pow(1 + interestRate, daysToLoanEnd * -1))
  const fvOfOriginalBalance =
    startValue * Math.pow(1 + interestRate, daysSinceStart)
  const fvOfAnnuity =
    repayment *
    ((Math.pow(1 + interestRate, daysSinceStart) - 1) / interestRate)
  return Math.round(fvOfOriginalBalance - fvOfAnnuity)
}
