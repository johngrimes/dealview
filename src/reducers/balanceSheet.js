export const InitialBalanceSheetState = {
  status: 'uninitialised',
  fresh: true,
  balanceSheet: {},
}

const BalanceSheetReducer = (state = InitialBalanceSheetState, action) => {
  switch (action.type) {
    case 'LOAD_BALANCE_SHEET_REQUEST':
    case 'UPDATE_BALANCE_SHEET_REQUEST':
      return {
        ...state,
        status: 'loading',
      }
    case 'LOAD_BALANCE_SHEET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        balanceSheet: action.balanceSheet,
      }
    case 'UPDATE_BALANCE_SHEET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        fresh: true,
        balanceSheet: action.balanceSheet,
      }
    case 'LOAD_BALANCE_SHEET_FAILURE':
    case 'UPDATE_BALANCE_SHEET_FAILURE':
      return {
        ...state,
        status: 'error',
        error: action.error,
      }
    case 'INVALIDATE_BALANCE_SHEET':
      return {
        ...state,
        fresh: false,
      }
    default:
      return state
  }
}

export default BalanceSheetReducer
