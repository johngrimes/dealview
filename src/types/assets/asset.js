// @flow

import moment from 'moment'

import { DateFormat } from 'types/commonTypes'
import type { Valuations, Valuation } from 'components/forms/ValuationsInput'

export type Asset = {
  type: 'RealEstate',
  instanceId: string,
  name: string,
  startDate: string,
  endDate?: string,
  valuations: Valuations
}

export type AssetWithId = Asset & { id: string }

export type AssetMap = { [id: string]: AssetWithId }

export const getValuationAtDate = (asset: { startDate: string, endDate?: string, valuations: Valuations },
                                   date: string): number => {
  const queryDate = moment(date)
  const startDate = moment(asset.startDate)
  const endDate = typeof asset.endDate === 'string'
    ? moment(asset.endDate)
    : undefined
  if (typeof asset.valuations === 'undefined' ||
      asset.valuations.length === 0 ||
      queryDate.isBefore(startDate) ||
      (endDate && queryDate.isAfter(endDate))) {
    return 0
  }
  const sortedValuations = asset.valuations.slice().sort(compareValuationsByDate)
  const idxFuture = sortedValuations.findIndex(v => {
    return moment(v.date, DateFormat).valueOf() > queryDate.valueOf()
  })
  const slicedValuations = idxFuture === -1
    ? sortedValuations
    : sortedValuations.slice(0, idxFuture)
  const amount = typeof slicedValuations[slicedValuations.length - 1] === 'undefined'
    ? 0
    : slicedValuations[slicedValuations.length - 1].amount
  return typeof amount === 'number' ? amount : 0
}

const compareValuationsByDate = (a: Valuation, b: Valuation): number => {
  const [ milliA, milliB ] = [ a, b ].map(v => { return v.date ? moment(v.date).valueOf() : 0 })
  return milliA - milliB
}
