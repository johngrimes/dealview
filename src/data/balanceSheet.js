import values from 'lodash.values'
import arrayHead from 'lodash.head'
import arrayTail from 'lodash.tail'
import moment from 'moment'

import { DateStorageFormat } from './commonTypes.js'
import { getValuationAtDate } from './asset.js'
import { getBalanceAtDate } from './loan.js'

export const balanceSheetOverTime = (
  assets,
  liabilities,
  startDate,
  endDate
) => {
  return Object.assign(
    {},
    ...calcBalanceSheet(assets, liabilities, startDate, endDate)
  )
}

function * calcBalanceSheet(assets, loans, date, endDate) {
  let nextDate = date
  while (!nextDate.isAfter(endDate)) {
    const totalAssets = sumAssetValueAtDate(values(assets), nextDate)
    const totalLiabilities = sumLoanBalanceAtDate(values(loans), nextDate)
    yield {
      [nextDate.format(DateStorageFormat)]: {
        totalAssets,
        totalLiabilities,
        equity: totalAssets - totalLiabilities,
      },
    }
    nextDate = moment(nextDate).add(1, 'days')
  }
}

const sumAssetValueAtDate = (assets, date) => {
  if (assets.length === 0) return 0
  const { head, tail } = { head: arrayHead(assets), tail: arrayTail(assets) }
  const headResult = getValuationAtDate(head, date)
  return headResult + sumAssetValueAtDate(tail, date)
}

const sumLoanBalanceAtDate = (loans, date) => {
  if (loans.length === 0) return 0
  const { head, tail } = { head: arrayHead(loans), tail: arrayTail(loans) }
  const headResult = getBalanceAtDate(head, date)
  return headResult + sumLoanBalanceAtDate(tail, date)
}
