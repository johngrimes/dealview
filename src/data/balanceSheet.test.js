import moment from 'moment'
import pick from 'lodash.pick'

import { balanceSheetOverTime } from './balanceSheet.js'
import { DateStorageFormat } from './commonTypes.js'
import { validAssetWithId1, validAssetWithId2 } from './fixtures/asset.js'
import { validLoanWithId1, validLoanWithId2 } from './fixtures/loan.js'

describe('balanceSheetOverTime', () => {
  const assets = {
    [validAssetWithId1.id]: validAssetWithId1,
    [validAssetWithId2.id]: validAssetWithId2,
  }
  const loans = {
    [validLoanWithId1.id]: validLoanWithId1,
    [validLoanWithId2.id]: validLoanWithId2,
  }

  it('should return correct results', () => {
    window.onmessage = () => {}
    const result = balanceSheetOverTime(
      assets,
      loans,
      moment('2019-08-20', DateStorageFormat),
      moment('2025-01-05', DateStorageFormat)
    )
    const prunedResult = pick(
      result,
      '2019-08-20',
      '2019-08-21',
      '2019-08-24',
      '2019-08-25',
      '2019-08-26',
      '2021-05-09',
      '2021-05-10',
      '2021-05-11',
      '2024-12-31',
      '2025-01-01',
      '2025-01-02',
      '2025-01-05'
    )
    expect(prunedResult).toMatchSnapshot()
  })
})
