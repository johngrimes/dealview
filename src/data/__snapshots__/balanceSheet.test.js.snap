// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balanceSheetOverTime should return correct results 1`] = `
Object {
  "2019-08-20": Object {
    "equity": 849754,
    "totalAssets": 1466122,
    "totalLiabilities": 616368,
  },
  "2019-08-21": Object {
    "equity": 849990,
    "totalAssets": 1466282,
    "totalLiabilities": 616292,
  },
  "2019-08-24": Object {
    "equity": 850700,
    "totalAssets": 1466765,
    "totalLiabilities": 616065,
  },
  "2019-08-25": Object {
    "equity": 850935,
    "totalAssets": 1466925,
    "totalLiabilities": 615990,
  },
  "2019-08-26": Object {
    "equity": 253470,
    "totalAssets": 869384,
    "totalLiabilities": 615914,
  },
  "2021-05-09": Object {
    "equity": 363462,
    "totalAssets": 930708,
    "totalLiabilities": 567246,
  },
  "2021-05-10": Object {
    "equity": 363646,
    "totalAssets": 930810,
    "totalLiabilities": 567164,
  },
  "2021-05-11": Object {
    "equity": 363829,
    "totalAssets": 930912,
    "totalLiabilities": 567083,
  },
  "2024-12-31": Object {
    "equity": 625809,
    "totalAssets": 1076974,
    "totalLiabilities": 451165,
  },
  "2025-01-01": Object {
    "equity": 626021,
    "totalAssets": 1077092,
    "totalLiabilities": 451071,
  },
  "2025-01-02": Object {
    "equity": 626233,
    "totalAssets": 1077210,
    "totalLiabilities": 450977,
  },
  "2025-01-05": Object {
    "equity": 626867,
    "totalAssets": 1077564,
    "totalLiabilities": 450697,
  },
}
`;
